jwst/ramp_fitting/ramp_fit.py

```diff
@@ -247,53 +254,71 @@ def ols_ramp_fit(model, buffsize, save_opt, readnoise_model, gain_model,
 
             data_sect = model.get_section('data')[num_int, :, rlo:rhi, :]
 
-            # skip data section if it is all NaNs
+            # Skip data section if it is all NaNs
             if  np.all(np.isnan( data_sect)):
                 log.error('Current data section is all nans, so not processing the section.')
-                break   
-
-            # Poisson noise only; read noise only
-            var_p_2d = np.zeros((max_seg, (rhi - rlo) * imshape[-1]),
-                                dtype=np.float32)
-            var_r_2d = np.zeros((max_seg, (rhi - rlo) * imshape[-1]),
-                                dtype=np.float32)
+                break
 
-            # first frame section for 1st read of current integration
+            # first frame section for 1st group of current integration
             ff_sect = model.get_section('data')[ num_int, 0, rlo:rhi, :].\
                 astype(np.float32)
-            # get appropriate sections
+
+            # Get appropriate sections
             gdq_sect = gdq_cube[num_int, :, rlo:rhi, :]
             rn_sect = readnoise_2d[rlo:rhi, :]
             gain_sect = gain_2d[rlo:rhi, :]
 
-            t_err_cube, t_dq_cube, m_by_var, inv_var, opt_res, f_max_seg = \
-                 calc_slope(data_sect, gdq_sect, frame_time, opt_res,
-                     rn_sect, gain_sect, max_seg, ngroups, weighting, f_max_seg,
-                     var_p_2d, var_r_2d)
+            # Reset all saturated groups in the input data array to NaN
+            where_sat = np.where( np.bitwise_and(gdq_sect,
+                                  dqflags.group['SATURATED']) != 0)
+            data_sect[ where_sat ] = np.NaN
+
+            # Compute the first differences of all groups
+            first_diffs_sect = np.diff(data_sect, axis=0)
+
+            # If the dataset has only 1 group/integ, assume the 'previous group'
+            #   is all zeros, so just use data as the difference
+            if (first_diffs_sect.shape[0] == 0):
+                first_diffs_sect = data_sect[0,:,:]
+                first_diffs_sect = first_diffs_sect[np.newaxis,:]
+
+            # Similarly, for datasets having >1 group/integ and having
+            #  single-group segments, just use the data as the difference
+            wh_nan = np.where( np.isnan( first_diffs_sect[0,:,:]) )
+            if (len(wh_nan[0]) > 0):
+                first_diffs_sect[0,:,:][ wh_nan ] = data_sect[0,:,:][ wh_nan ]
+
+            # Mask all the first differences that are affected by a CR, starting 
+            #  at group 1.  The purpose of starting at index 1 is to shift all the
+            #  indexes down by 1, so they line up with the indexes in first_diffs.
+            i_group, i_yy, i_xx, = np.where( np.bitwise_and(gdq_sect[:,:,1:],
```

Comment:
Shouldn't the slice be `gdq_sect[1:, :, :]`?  And if so, won't that be a problem in the next line for data with only one group when you index with `i_group-1`?

Reply:
Yes, and fixing this uncovered a semi-related bug (now fixed).